### ISSUE TITLE
IndexedDB not working

### ISSUE THREAD

**Issue by cpgb85** [agershun/alasql#1775]
I'm using the example to use indexeddb because I have a larger amount of data to store. The official example creates the IndexedDB but any sql statement does not work. I can fetch all rows fine, but once i use "SELECT COLUMN * FROM cities WHERE population > 100000 ORDER BY city DESC" it returns nothing.

Is there any more resources besides this single example to use indexeddb?

**Comment by cpgb85** [comment]
I have my code this far, but it doesn't actually query any columns: 

``` 
        await alasql(['CREATE INDEXEDDB DATABASE IF NOT EXISTS ixbase']);
        await alasql(['ATTACH INDEXEDDB DATABASE ixbase']);
        await alasql(['USE ixbase'])

        await alasql.promise("CREATE TABLE if not exists cities (population number, city string)"); 
        await alasql.promise("select * into cities from ?", [[{city:"Redmond", population:57530},{city:"Atlanta",population:447841},{city:"San Fracisco", population:837442}]]);
        let res = await alasql.promise(`select * from cities c WHERE c.population > 80000`);
        console.log(res)
```

**Comment by mathiasrw** [comment]
I see. 

I made this jsfilddle

https://jsfiddle.net/3hueqz1s/

I can confirm that if you comment out the 3 lines connecting to indexedDb then it works. 

It looks like 
<img width="324" alt="image" src="https://github.com/AlaSQL/alasql/assets/1063454/0928d26e-d6fc-4772-a357-f0102d62242a">
is returning before we got data persisted in indexedDB

**Comment by TheBeard30** [comment]
> I see.
> 
> I made this jsfilddle
> 
> https://jsfiddle.net/3hueqz1s/
> 
> I can confirm that if you comment out the 3 lines connecting to indexedDb then it works.
> 
> It looks like <img alt="image" width="324" src="https://user-images.githubusercontent.com/1063454/261054721-0928d26e-d6fc-4772-a357-f0102d62242a.png"> is returning before we got data persisted in indexedDB

So which version is available about IndexedDB？

**Comment by TheBeard30** [comment]
> I see.
> 
> I made this jsfilddle
> 
> https://jsfiddle.net/3hueqz1s/
> 
> I can confirm that if you comment out the 3 lines connecting to indexedDb then it works.
> 
> It looks like <img alt="image" width="324" src="https://user-images.githubusercontent.com/1063454/261054721-0928d26e-d6fc-4772-a357-f0102d62242a.png"> is returning before we got data persisted in indexedDB

when sql statement have 'WHERE CONDITIONS'，the callback function always have empty array data in this jsfilddle demo @mathiasrw

**Comment by eguchi-naoya** [comment]
Sorry to ask, if you are busy, but i just want to know, Is this bug will be fixed one day?

v4.1.2: OK
        await alasql.promise("ATTACH INDEXEDDB DATABASE ExampleComDatabase;");
        await alasql.promise("USE DATABASE ExampleComDatabase;");
        const result = await alasql.promise("SELECT * FROM customer WHERE customerSid=1");
        console.log(result);//OK: return right data

v4.1.3～v4.1.10: NG
        await alasql.promise("ATTACH INDEXEDDB DATABASE ExampleComDatabase;");
        await alasql.promise("USE DATABASE ExampleComDatabase;");
        const result = await alasql.promise("SELECT * FROM customer WHERE customerSid=1");
        console.log(result); //NG: return empty array.
 
        //Without WHERE is working, I think specify any column in SQL with INDEXEDDB is not working, since v4.1.3's fix.
        const result2 = await alasql.promise("SELECT * FROM customer");
        console.log(result2); //OK: return array.

**Comment by cpgb85** [comment]
> I see.
> 
> I made this jsfilddle
> 
> https://jsfiddle.net/3hueqz1s/
> 
> I can confirm that if you comment out the 3 lines connecting to indexedDb then it works.
> 
> It looks like <img alt="image" width="324" src="https://user-images.githubusercontent.com/1063454/261054721-0928d26e-d6fc-4772-a357-f0102d62242a.png"> is returning before we got data persisted in indexedDB

that literally defeats the entire purpose. I want to use IndexedDB to store large amounts of data.

**Comment by mathiasrw** [comment]
@eguchi-naoya super happy to merge a pull request when one is ready.

**Comment by eguchi-naoya** [comment]
@mathiasrw 
Thank you for your reply.
Unfortunately I can't help this project, because I'm not familiar with Typescript.

I decided to use this wonderful solution because 
This was an only and excellent solution that allowed me to manipulate indexdb using SQL
, which I was familiar with in Java(20 years) when I was creating a client-side es6 pure javscript application as a hobby.

Once If I become good at Typescript, I would like to support this wonderful project. until then I will be happy to use v4.1.2.
Thanks.

**Comment by cpgb85** [comment]
> @mathiasrw Thank you for your reply. Unfortunately I can't help this project, because I'm not familiar with Typescript.
> 
> I decided to use this wonderful solution because This was an only and excellent solution that allowed me to manipulate indexdb using SQL , which I was familiar with in Java(20 years) when I was creating a client-side es6 pure javscript application as a hobby.
> 
> Once If I become good at Typescript, I would like to support this wonderful project. until then I will be happy to use v4.1.2. Thanks.

Can you give example code of how you used indexeddb?

**Comment by eguchi-naoya** [comment]
> > @mathiasrw Thank you for your reply. Unfortunately I can't help this project, because I'm not familiar with Typescript.
> > I decided to use this wonderful solution because This was an only and excellent solution that allowed me to manipulate indexdb using SQL , which I was familiar with in Java(20 years) when I was creating a client-side es6 pure javscript application as a hobby.
> > Once If I become good at Typescript, I would like to support this wonderful project. until then I will be happy to use v4.1.2. Thanks.
> 
> Can you give example code of how you used indexeddb?

Try this html example in your browser.
※specify version for script tag 'alasql@4.1.2' is important

```html
<html>

<head>
    <script type="text/javascript" src="https://cdn.jsdelivr.net/npm/alasql@4.1.2/dist/alasql.min.js"></script>
    <script type="module">
        await alasql(['CREATE INDEXEDDB DATABASE IF NOT EXISTS ixbase']);
        await alasql(['ATTACH INDEXEDDB DATABASE ixbase']);
        await alasql(['USE ixbase'])
        await alasql.promise("CREATE TABLE if not exists cities (population number, city string)");
        //delete previous run inserted data to be always same select results
        await alasql.promise("DELETE FROM cities WHERE city IN('Redmond', 'Atlanta', 'San Fracisco')");
        //insert data for select test
        await alasql.promise("SELECT * INTO cities FROM ?", [
            [{
                city: "Redmond",
                population: 57530
            }, {
                city: "Atlanta",
                population: 447841
            }, {
                city: "San Fracisco",
                population: 837442
            }]
        ]);
        let res = await alasql.promise(`SELECT * FROM cities c WHERE c.population > 80000`);
        //see the browser console for result
        console.log(res);
    </script>
</head>

<body></body>

</html>
```

**Comment by sytone** [comment]
ok, found the issue for this one. There was a change to the IndexedDB code which introduced a change around the key value.

It is checking to see if the cursor is a object by `typeof cursor === Object`. I believe it should be `typeof cursor === "object"` looking at documentation.

https://github.com/vishal6557/alasql/blob/44931206290cd2526835252aee16aaff24fcc2b0/src/91indexeddb.js#L342C1-L345C23

This change was to fix #1556 but it broke the ability to normally query, the tests added did not check the other path so when a WHERE query is run against the values there are no actual columns as it is always in the form of { "KEYNUMBER": { { "col1": "val1", "col2": "val2" } } }

I think I know the code changes to make to fix it but will not have time this week to make the changes.

**Comment by mathiasrw** [comment]
@sytone Amazing! Love it!

Let me know if you need help.

**Comment by sytone** [comment]
I think there may be a issue with all tests running. Looking at the test output there are tests that are not running that have a object check in them. Is this expected?

**Comment by sytone** [comment]
@mathiasrw are you able to check to see if all tests are actually running as expected?

**Comment by mathiasrw** [comment]
@sytone Yep, all good good

https://github.com/AlaSQL/alasql/actions/runs/8839168786/job/24271937513?pr=1914

**Comment by mathiasrw** [comment]
Fix released as part of v4.3.3
